cm_data: Handled fi_shutdown event and added code to synchronize via socket communication

shantonu opened the pull request:
- According to manpage <code>fi_shutdown</code> generates an FI_SHUTDOWN event. Added <code>fi_eq_sread</code> to handle the event.
- Added a memset before eq_sread after fi_listen
- Used FT_EQ_ERR macro for fi_eq_readerr case
- Added cm_data into runfabtests.sh
- Moved sockets communication code from complex folder to <code>common/shared.c</code> and <code>include/shared.h</code>
- Renamed ft_fw_* functions as ft_sock_* functions for clarification
- Added sockets synchronization code to address Travis failures
- Added an entry for cm_data in the manpage

@jithinjosepkl @bturrubiates 
Signed-off-by: Shantonu Hossain <shantonu.hossain@intel.com>

shantonu commented:
Not sure why Travis-CI was not run. @shefty, do you have any idea?

jithinjosepkl commented on simple/cm_data.c:
```diff
@@ -191,6 +191,19 @@ static int server_accept(size_t paramlen)
 
 	fi_shutdown(ep, 0);
+	memset(entry, 0, sizeof(*entry) + paramlen);
+	ret = fi_eq_sread(eq, &event, entry, sizeof(*entry) + paramlen, -1, 0);
```
nit: FI_SHUTDOWN event will not have cm-data. It is ok to have a larger buffer for eq_sread, but it will be better to avoid any confusion here.

shantonu replied:
Good catch! Fixed it.

jithinjosepkl commented:
Can you please re-trigger the travis-CI, so that it runs with the updated code?

bturrubiates commented:
> Can you please re-trigger the travis-CI, so that it runs with the updated code?

I re-triggered it using the Travis interface.

shefty commented on simple/cm_data.c:
```diff
@@ -191,6 +191,19 @@ static int server_accept(size_t paramlen)
 
 	fi_shutdown(ep, 0);
+	memset(entry, 0, sizeof(*entry) + paramlen);
+	ret = fi_eq_sread(eq, &event, entry, sizeof(*entry), -1, 0);
```
It looks odd seeing memset use sizeof(*entry) + paramlen, but sread only use sizeof(*entry)

shantonu replied:
ok, will fix it.

shefty commented on simple/cm_data.c:
```diff
@@ -288,6 +301,19 @@ static int client_expect_accept(size_t paramlen)
+			entry->fid, ep);
+		ret = -FI_EOTHER;
+		return ret;
+	}
```
Can we move these two blocks of code into a function, rather than repeating them?

shantonu replied:
Done!

jithinjosepkl commented:
@bturrubiates - thanks for triggering it. Btw, how did you trigger it? I couldn't any option in travis page.

Looks like some failures, need to check this.

bturrubiates commented:
> @bturrubiates - thanks for triggering it. Btw, how did you trigger it? I couldn't any option in travis page.

It's in the top right, [this](http://s3.amazonaws.com/awesome_screenshot/9972367?AWSAccessKeyId=0R7FMW7AXRVCYMAPTPR2&Expires=1465506022&Signature=veWB0KDYAoJlmNrx0tpPhvbDn5U%3D) image points to it with a red arrow.

jithinjosepkl commented:
Thanks, thats useful info :)

shefty commented on simple/cm_data.c:
```diff
@@ -46,6 +46,26 @@ static size_t cm_data_size;
+	int ret;
+	uint32_t event;
+
+	memset(entry, 0, sizeof(*entry));
```
I don't see where entry is ever freed in the code.

shantonu replied:
Yes, will add that.

shefty commented:
Please see the travis failures, which look legitimate.

shefty commented on simple/cm_data.c:
```diff
@@ -356,7 +383,8 @@ static int run(void)
 		}
 	}
-
+err:
```
please rename this label to 'out' as non-error paths go here as well.

shantonu replied:
Ok!

jithinjosepkl commented:
I think this one is ready to merge, cm_data test seems to be running fine.

jithinjosepkl commented:
Hmm, runs into an error after rebase. Will need to check this.

shefty commented:
There are open comments against this still.  Maybe we're missing a push update?

shantonu commented:
There was a minor open comment. I pushed it but it still fails for clang. From the build log, I see last time it passed with libfabric PR #2174. There was no changes in fabtests that can impact cm_data since then. So I am guessing either changes to sockets provider after PR 2174 introduced a bug in cm_data or there are some issue in the latest sockets provider PRs.

jithinjosepkl commented:
@bturrubiates, @shantonu - The travis failures are because of a race between the FI_CONNREQ and FI_SHUTDOWN events. When the failure happens, FI_CONNREQ from client for the next iteration comes before the FI_SHUTDOWN event. So, the server expects FI_SHUTDOWN event, but gets FI_CONNREQ with cm-data, and thus generates "buffer too small" error.

I think it is a valid case that the FI_CONNREQ comes before the FI_SHUTDOWN, because both client and server are calling fi_shutdown. I think the test should somehow be updated to handle this case. May be we need to add some sort of explicit synchronization or asynchronous handling of EQ events?

Any thoughts?

bturrubiates commented:
> I think it is a valid case that the FI_CONNREQ comes before the FI_SHUTDOWN, because both client and server are calling fi_shutdown. I think the test should somehow be updated to handle this case. May be we need to add some sort of explicit synchronization or asynchronous handling of EQ events?

We could add a sync point after the call to `read_shutdown_event()`.

shantonu commented:
We can't really do send/recv synchronization after fi_shutdown similar to <code>ft_sync()</code> in shared.c. @bturrubiates do you have any suggestions?

jithinjosepkl commented:
One option is to use a separate EP (or keep the first EP created) for sync msgs. Or we can add asynchronous EQ event handling at the server side.

jithinjosepkl commented:
@shantonu, @bturrubiates - Can we finalize this one? If there are no other thoughts, can we go ahead with one of these approaches?

bturrubiates commented:
> @shantonu, @bturrubiates - Can we finalize this one? If there are no other thoughts, can we go ahead with one of these approaches?

I'm looking at this right now and it seems to consistently fail on the 2nd iteration, is this the failure you were seeing?

jithinjosepkl commented:
Right, and the failure is because we are expecting a shutdown event, but it gets the conn-req event from client (from next iteration).

shantonu commented:
Just rebased the branch and tried with libfabric master. Works for me. The failure is random, I guess.

bturrubiates commented:
> Just rebased the branch and tried with libfabric master. Works for me. The failure is random, I guess.

Actually, that was my mistake. I think I was running an outdated version of this test. Seems to be working well for me now. Let's see what Travis thinks.

bturrubiates commented:
This will fail for us until I fix ofiwg/libfabric#946.

bturrubiates commented:
Actually, just realized we're not checking the return call from `fi_shutdown`. Maybe check that and make sure it is  `FI_SUCCESS`, before continuing.

bturrubiates commented:
Hmm, yeah the race is still there under Travis.

jithinjosepkl commented:
Yes, the race is because the server gets a new-conn-req, while waiting for the shutdown event. In the test, we don't sync between iterations. so the conn-req (from the next iteration) comes earlier than the local disconnect event.

bturrubiates commented:
> Yes, the race is because the server gets a new-conn-req, while waiting for the shutdown event. In the test, we don't sync between iterations. so the conn-req (from the next iteration) comes earlier than the local disconnect event.

I was playing around with it trying to use other sync points, but I think the best way to do this will just be to use another EP or another out of band mechanism like sockets.

bturrubiates commented:
I don't think asynchronous event handling will be very useful because at the time we get the connreq we haven't started listening with the new size.

jithinjosepkl commented:
Yes, we can use either a separate EP for synchronization or ```ft_fw_sync()``` that fi_ubertest uses.

bturrubiates commented:
> Yes, we can use either a separate EP for synchronization or ft_fw_sync() that fi_ubertest uses.

`ft_fw_sync` is only available in the complex subfolder, I guess we could always move it up a directory. It might be easier to just use a second EP.

shantonu commented:
I was wondering if adding a second ep would complicate the simple test. But I'm fine with either option.

bturrubiates commented:
> I was wondering if adding a second ep would complicate the simple test. But I'm fine with either option.

We could use sockets and add it to the common code.

shantonu commented:
I'll update the PR with sockets code.

shantonu commented:
Sorry for the long PR. 
- Moved sockets communication code from complex folder to <code>common/shared.c</code> and <code>include/shared.h</code>
- Renamed ft_fw_* functions as ft_sock_* functions for clarification
- Added sockets synchronization code to address Travis failures

@jithinjosepkl @bturrubiates @a-ilango, can you please review? Travis seems to work fine.

jithinjosepkl commented:
Thanks @shantonu. Should we split this into separate commits?

jithinjosepkl commented on simple/cm_data.c:
```diff
@@ -371,8 +423,11 @@ int main(int argc, char **argv)
 		return EXIT_FAILURE;
 
-	while ((op = getopt(argc, argv, "h" ADDR_OPTS INFO_OPTS)) != -1) {
+	while ((op = getopt(argc, argv, "q:h" ADDR_OPTS INFO_OPTS)) != -1) {
```
Do we need to do anythign else for showing up in help text?

shantonu replied:
Oh, I forgot to add it to the help text.

shantonu commented:
I was thinking of separate commits but without this refactoring the PR will fail on Travis.

jithinjosepkl commented:
Hmm, it can be in the same PR, but different commits will be helpful.

shantonu commented:
I think Travis runs for each commit. So the first commit will fail.

jithinjosepkl commented:
Afaik, It is per PR, not per commit. (https://travis-ci.org/ofiwg/fabtests/pull_requests).

shantonu commented:
Hmm, I thought the green check mark after the commit means they are tested individually. Anyway, I have created three commits for better understanding.

jsquyres commented:
@shantonu There's currently 2 CI tests on OFIWG PRs:

1. The signed off checker.  This one runs on every single commit, and will show a green check or red X for each commit.  We decided to do it this way (vs. only indicating a green check/red X on the last/most recent commit) to avoid potential confusion.  E.g., if there's N commits, and *1* of them is not signed off, but the *last* commit is signed off and still gets a red X anyway.
1. Travis.  Travis will only run on the *last* commit in an atomic push sequence.  Hence, if you push 5 commits simultaneously, Travis will test only the *last* commit.  If you then push more commits after those 5, Travis will then test the (new) last commit.

Make sense?

If you hover over each of the green check marks, you'll see that the signed off checker ran on all of them, but the Travis check only ran on the last one.

More specifically:

1. The Signed Off Checker is intentionally checking *every* commit.
1. Travis is intentionally only checking the *last* commit.

shantonu commented:
Thanks for the clarification!

jithinjosepkl commented on simple/cm_data.c:
```diff
@@ -382,8 +386,28 @@ static int run(void)
-err:
+
+err1:  ft_sock_shutdown(sock);
+err2:
```
We can use a single label here.

shantonu replied:
There are lines after err2. Added a newline so that it is shown in the diff.

jithinjosepkl commented on simple/cm_data.c:
```diff
@@ -382,8 +386,28 @@ static int run(void)
+			ret = ft_sock_sync(0);
+			if (ret)
+				goto err1;
+		}
```
Minor suggestion: We can keep it as a single loop, and with the connect/accept calls before the loop. That we we don't need to have the loop code repeated.

shantonu replied:
Sure, done!

jithinjosepkl commented:
Thanks @shantonu 👍

jithinjosepkl commented on simple/cm_data.c:
```diff
@@ -357,6 +403,10 @@ static int run(void)
+	ft_sock_shutdown(sock);
+err2:
+	free(entry);
 	return ret;
```
👍

shantonu commented:
If there's no other comments, can we merge this? @jithinjosepkl @bturrubiates

jithinjosepkl commented:
👍

bturrubiates commented:
I think I'm probably good with this 👍
